### PR TITLE
[Run2_2017] Store initial partons in gen particles list.

### PR DESCRIPTION
**Pull request by aperloff**
This PR satisfies #536 by saving the initial partons in the gen particles list.

### CODE REVIEW

**Comment by aperloff** [comment]
This increases the size of the ntuple by ~0.66% for a TTJets sample using the V18e settings.

I looked at the output of the following samples:
```bash
Summer16v3.TTJets_SingleLeptFromT_TuneCUETP8M1_13TeV-madgraphMLM-pythia8
Summer16v3.ZJetsToNuNu_HT-100To200_13TeV-madgraph
Summer16v3.QCD_HT300to500_TuneCUETP8M1_13TeV-madgraphMLM-pythia8
Summer16v3.QCD_Pt_470to600_TuneCUETP8M1_13TeV_pythia8
Summer16v3.SMS-T5qqqqZH-mGluino-1000to2500_TuneCUETP8M1_13TeV-madgraphMLM-pythia8
PrivateSamples.EMJ_2016_mMed-1000_mDark-20_ctau-1000_unflavored-down
```
In most of them this this change did exactly what was expected, increase the gen particle count by 2. In the two QCD samples there are occasional events where only one parton is added. The table below shows an example of this.

```bash
root [3] PreSelection->Scan("GenParticles_PdgId:GenParticles_ParentId:GenParticles_ParentIdx:GenParticles_Status")
***********************************************************************
*    Row   * Instance * GenPartic * GenPartic * GenPartic * GenPartic *
***********************************************************************
...
*        6 *        0 *        21 *      2212 *        -1 *        21 *
*        6 *        1 *        21 *      2212 *        -1 *        21 *
*        7 *        0 *        21 *      2212 *        -1 *        21 *
*        8 *        0 *         1 *      2212 *        -1 *        21 *
*        8 *        1 *        -1 *      2212 *        -1 *        21 *
```

File: Utils/src/GenParticlesProducer.cc [hunk 1]
```diff
@@ -302,42 +304,66 @@ void GenParticlesProducer::storeMinimal(const edm::Handle< edm::View<reco::GenPa
...
         // This is to protect against parts of a chain which may be saved, but here we are looking at the first copy
-        if(stored_particles_ref.find(&iPart)!=stored_particles_ref.end()) continue;
+        if(stored_particles_ref.find(&particle)!=stored_particles_ref.end()) continue;
+
+        // Store the initial partons
+        auto nmothers = particle.numberOfMothers();
+        if(nmothers == 1) {
+           auto mother = particle.mother(0);
+           if(mother->pdgId() == proton &&
+              std::binary_search(proton_status_codes.begin(), proton_status_codes.end(), mother->status()) &&
```
Comment (kpedro88) [inline Utils/src/GenParticlesProducer.cc]: as long as there's only one status code, `binary_search` is probably overkill - just compare to it directly

File: Utils/src/GenParticlesProducer.cc [hunk 2]
```diff
@@ -302,42 +304,66 @@ void GenParticlesProducer::storeMinimal(const edm::Handle< edm::View<reco::GenPa
...
+              (mother == &(*genPartCands->begin()) || mother == &(*(genPartCands->begin()+1))) &&
+              particle.status()==21) {
```
Comment (kpedro88) [inline Utils/src/GenParticlesProducer.cc]: maybe worth denoting 21 as a constant like `initial_parton_status_code`

File: Utils/src/GenParticlesProducer.cc [hunk 3]
```diff
@@ -302,42 +304,66 @@ void GenParticlesProducer::storeMinimal(const edm::Handle< edm::View<reco::GenPa
...
-        if(stored_particles_ref.find(&iPart)!=stored_particles_ref.end()) continue;
+        if(stored_particles_ref.find(&particle)!=stored_particles_ref.end()) continue;
+
+        // Store the initial partons
+        auto nmothers = particle.numberOfMothers();
+        if(nmothers == 1) {
+           auto mother = particle.mother(0);
+           if(mother->pdgId() == proton &&
+              std::binary_search(proton_status_codes.begin(), proton_status_codes.end(), mother->status()) &&
+              (mother == &(*genPartCands->begin()) || mother == &(*(genPartCands->begin()+1))) &&
```
Comment (kpedro88) [inline Utils/src/GenParticlesProducer.cc]: could keep the comparison pointers outside the loop:
```cpp
std::array<const reco::Candidate*, 2> possible_mothers{&(*genPartCands->begin()), &(*(genPartCands->begin()+1)))};
```
and then compare to `possible_mothers[0]` and `possible_mothers[1]` inside the loop (avoid repeated creation of iterators, dereferencing, etc.)

Comment (aperloff) [reply Utils/src/GenParticlesProducer.cc]: Done

File: Utils/src/GenParticlesProducer.cc (hunk 2 above)
Comment (aperloff) [reply Utils/src/GenParticlesProducer.cc]: Done

File: Utils/src/GenParticlesProducer.cc (hunk 1 above)
Comment (aperloff) [reply Utils/src/GenParticlesProducer.cc]: Done

**Pull request merged**